File: app/src/main/java/jp/hazuki/yuzubrowser/action/item/ShowSearchBoxAction.java
/*
 * Copyright (C) 2017 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.action.item;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Parcel;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.CheckBox;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.action.SingleAction;
import jp.hazuki.yuzubrowser.action.view.ActionActivity;
import jp.hazuki.yuzubrowser.utils.app.StartActivityInfo;

public class ShowSearchBoxAction extends SingleAction implements Parcelable {
    private static final String FIELD_OPEN_NEW_TAB = "0";
    private static final String FIELD_REVERSE = "1";

    private boolean openNewTab;
    private boolean reverse;

    public ShowSearchBoxAction(int id, JsonParser parser) throws IOException {
        super(id);
        if (parser != null) {
            if (parser.nextToken() != JsonToken.START_OBJECT) return;
            while (parser.nextToken() != JsonToken.END_OBJECT) {
                if (FIELD_OPEN_NEW_TAB.equals(parser.getCurrentName())) {
                    if (parser.nextValue().isBoolean())
                        openNewTab = parser.getBooleanValue();
                }
                if (FIELD_REVERSE.equals(parser.getCurrentName())) {
                    if (parser.nextValue().isBoolean())
                        reverse = parser.getBooleanValue();
                }

                if (parser.getCurrentToken() == JsonToken.START_OBJECT
                        || parser.getCurrentToken() == JsonToken.START_ARRAY) {
                    parser.skipChildren();
                }
            }
        }
    }

    @Override
    public void writeIdAndData(JsonGenerator generator) throws IOException {
        generator.writeNumber(id);
        generator.writeStartObject();
        generator.writeBooleanField(FIELD_OPEN_NEW_TAB, openNewTab);
        generator.writeBooleanField(FIELD_REVERSE, reverse);
        generator.writeEndObject();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeInt(openNewTab ? 1 : 0);
        dest.writeInt(reverse ? 1 : 0);
    }

    protected ShowSearchBoxAction(Parcel source) {
        super(source);
        openNewTab = source.readInt() != 0;
        reverse = source.readInt() != 0;
    }

    public static final Creator<ShowSearchBoxAction> CREATOR = new Creator<ShowSearchBoxAction>() {
        @Override
        public ShowSearchBoxAction createFromParcel(Parcel source) {
            return new ShowSearchBoxAction(source);
        }

        @Override
        public ShowSearchBoxAction[] newArray(int size) {
            return new ShowSearchBoxAction[size];
        }
    };

    @Override
    public StartActivityInfo showSubPreference(ActionActivity context) {
        View view = LayoutInflater.from(context).inflate(R.layout.action_show_search_box, null);
        final CheckBox checkBox = view.findViewById(R.id.checkBox);
        final CheckBox bottom = view.findViewById(R.id.checkBox2);
        checkBox.setChecked(openNewTab);
        bottom.setChecked(reverse);
        new AlertDialog.Builder(context)
                .setTitle(R.string.action_settings)
                .setView(view)
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        openNewTab = checkBox.isChecked();
                        reverse = bottom.isChecked();
                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();
        return null;
    }

    public boolean isOpenNewTab() {
        return openNewTab;
    }

    public boolean isReverse() {
        return reverse;
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/action/item/ShowSearchBoxAction.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.action.item

import android.app.AlertDialog
import android.os.Parcel
import android.os.Parcelable
import android.view.View
import android.widget.ArrayAdapter
import android.widget.CheckBox
import android.widget.Spinner
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.legacy.R
import jp.hazuki.yuzubrowser.legacy.action.SingleAction
import jp.hazuki.yuzubrowser.legacy.action.view.ActionActivity
import jp.hazuki.yuzubrowser.ui.app.StartActivityInfo
import java.io.IOException

class ShowSearchBoxAction : SingleAction, Parcelable {

    var openNewTabMode = 0
        private set
    var isReverse: Boolean = false
        private set

    @Throws(IOException::class)
    constructor(id: Int, reader: JsonReader?) : super(id) {
        if (reader != null) {
            if (reader.peek() != JsonReader.Token.BEGIN_OBJECT) return
            reader.beginObject()
            while (reader.hasNext()) {
                if (reader.peek() != JsonReader.Token.NAME) return
                when (reader.nextName()) {
                    FIELD_OPEN_NEW_TAB -> {
                        if (reader.peek() != JsonReader.Token.BOOLEAN) return
                        openNewTabMode =
                            if (reader.nextBoolean()) TAB_TYPE_NEW_TAB else TAB_TYPE_CURRENT
                    }
                    FIELD_REVERSE -> {
                        if (reader.peek() != JsonReader.Token.BOOLEAN) return
                        isReverse = reader.nextBoolean()
                    }
                    FIELD_TAB_TYPE -> {
                        if (reader.peek() != JsonReader.Token.NUMBER) return
                        openNewTabMode = reader.nextInt()
                    }
                    else -> reader.skipValue()
                }
            }
            reader.endObject()
        }
    }

    @Throws(IOException::class)
    override fun writeIdAndData(writer: JsonWriter) {
        writer.value(id)
        writer.beginObject()
        writer.name(FIELD_TAB_TYPE)
        writer.value(openNewTabMode)
        writer.name(FIELD_REVERSE)
        writer.value(isReverse)
        writer.endObject()
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeInt(openNewTabMode)
        dest.writeInt(if (isReverse) 1 else 0)
    }

    private constructor(source: Parcel) : super(source) {
        openNewTabMode = source.readInt()
        isReverse = source.readInt() != 0
    }

    override fun showSubPreference(context: ActionActivity): StartActivityInfo? {
        val view = View.inflate(context, R.layout.action_show_search_box, null)
        val spinner: Spinner = view.findViewById(R.id.tabTypeSpinner)
        val bottom = view.findViewById<CheckBox>(R.id.checkBox2)
        ArrayAdapter.createFromResource(
            context,
            R.array.action_show_search_list,
            android.R.layout.simple_spinner_item
        ).also { adapter ->
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            spinner.adapter = adapter
            spinner.setSelection(openNewTabMode)
        }

        bottom.isChecked = isReverse
        AlertDialog.Builder(context)
            .setTitle(R.string.action_settings)
            .setView(view)
            .setPositiveButton(android.R.string.ok) { _, _ ->
                openNewTabMode = spinner.selectedItemPosition
                isReverse = bottom.isChecked
            }
            .setNegativeButton(android.R.string.cancel, null)
                .show()
        return null
    }

    companion object {
        private const val FIELD_OPEN_NEW_TAB = "0"
        private const val FIELD_REVERSE = "1"
        private const val FIELD_TAB_TYPE = "2"

        const val TAB_TYPE_CURRENT = 0
        const val TAB_TYPE_NEW_TAB = 1
        const val TAB_TYPE_NEW_RIGHT_TAB = 2

        @JvmField
        val CREATOR: Parcelable.Creator<ShowSearchBoxAction> = object : Parcelable.Creator<ShowSearchBoxAction> {
            override fun createFromParcel(source: Parcel): ShowSearchBoxAction {
                return ShowSearchBoxAction(source)
            }

            override fun newArray(size: Int): Array<ShowSearchBoxAction?> {
                return arrayOfNulls(size)
            }
        }
    }
}
